timemachine/cpp/src/context.cu

```diff
@@ -50,6 +64,342 @@ Context::~Context() {
+        nb_pot) {
+        return nb_pot->get_cutoff();
+    } else if (std::shared_ptr<NonbondedAllPairs<float, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<float, false>>(pot);
+               nb_pot) {
+        return nb_pot->get_cutoff();
+    } else if (std::shared_ptr<NonbondedAllPairs<double, true>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, true>>(pot);
+               nb_pot) {
+        return nb_pot->get_cutoff();
+    } else if (std::shared_ptr<NonbondedAllPairs<double, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, false>>(pot);
+               nb_pot) {
+        return nb_pot->get_cutoff();
+    } else {
+        throw std::runtime_error("Unable to cast potential to NonbondedAllPairs");
+    }
+}
+
+bool is_nonbonded_potential(std::shared_ptr<Potential> pot) {
+    if (std::shared_ptr<NonbondedAllPairs<float, true>> nb_pot =
+            std::dynamic_pointer_cast<NonbondedAllPairs<float, true>>(pot);
+        nb_pot) {
+        return true;
+    } else if (std::shared_ptr<NonbondedAllPairs<float, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<float, false>>(pot);
+               nb_pot) {
+        return true;
+    } else if (std::shared_ptr<NonbondedAllPairs<double, true>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, true>>(pot);
+               nb_pot) {
+        return true;
+    } else if (std::shared_ptr<NonbondedAllPairs<double, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, false>>(pot);
+               nb_pot) {
+        return true;
+    }
+    return false;
+}
+
+void set_nonbonded_potential_idxs(
+    std::shared_ptr<Potential> pot, const int num_idxs, const unsigned int *d_idxs, const cudaStream_t stream) {
+    if (std::shared_ptr<NonbondedAllPairs<float, true>> nb_pot =
+            std::dynamic_pointer_cast<NonbondedAllPairs<float, true>>(pot);
+        nb_pot) {
+        nb_pot->set_atom_idxs_device(num_idxs, d_idxs, stream);
+    } else if (std::shared_ptr<NonbondedAllPairs<float, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<float, false>>(pot);
+               nb_pot) {
+        nb_pot->set_atom_idxs_device(num_idxs, d_idxs, stream);
+    } else if (std::shared_ptr<NonbondedAllPairs<double, true>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, true>>(pot);
+               nb_pot) {
+        nb_pot->set_atom_idxs_device(num_idxs, d_idxs, stream);
+    } else if (std::shared_ptr<NonbondedAllPairs<double, false>> nb_pot =
+                   std::dynamic_pointer_cast<NonbondedAllPairs<double, false>>(pot);
+               nb_pot) {
+        nb_pot->set_atom_idxs_device(num_idxs, d_idxs, stream);
+    } else {
+        throw std::runtime_error("unable to cast potential to NonbondedAllPairs");
+    }
+}
```

Comment:
That this verbose casting and error checking is necessary makes me want to consider moving `local_md` into a subclass of Context with sigature like (Python pseudocode):
```
LocalMDContext(nonbonded: NonbondedAllPairs, others: List[Potential], ...)
```

Reply:
Refer to https://github.com/proteneer/timemachine/pull/806#discussion_r957379690